[ILM] Add date setting to calculate index age

andreidan opened the pull request:
Add the `index.lifecycle.origination_date` to allow users to configure a
custom date that'll be used to calculate the index age for the phase
transmissions (as opposed to the default index creation date).

This could be useful for users to create an index with an "older"
origination date when indexing old data.

Relates to #42449.

<!--
Thank you for your interest in and contributing to Elasticsearch! There
are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your
attention.
-->

- Have you signed the [contributor license agreement](https://www.elastic.co/contributor-agreement)?
- Have you followed the [contributor guidelines](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md)?
- If submitting code, have you built your formula locally prior to submission with `gradle check`?
- If submitting code, is your pull request against master? Unless there is a good reason otherwise, we prefer pull requests against master and will backport as needed.
- If submitting code, have you checked that your submission is for an [OS and architecture that we support](https://www.elastic.co/support/matrix#show_os)?
- If you are submitting this code for a class then read our [policy](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md#contributing-as-part-of-a-class) for that.

andreidan committed:
[ILM] Add date setting to calculate index age

Add the `index.lifecycle.origination_date` to allow users to configure a
custom date that'll be used to calculate the index age for the phase
transmissions (as opposed to the default index creation date).

This could be useful for users to create an index with an "older"
origination date when indexing old data.

Relates to #42449.

elasticmachine commented:
Pinging @elastic/es-core-features

andreidan commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleExecutionState.java:
```diff
@@ -259,6 +269,7 @@ public int hashCode() {
         private Long phaseTime;
         private Long actionTime;
         private Long stepTime;
+        private Long indexOriginationDate;
```
As the `index.creation_date` and `index.lifecycle.origination_date` settings have default values, would it be worth working with primitives here?

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -208,6 +208,8 @@ public void validate(Integer numRoutingShards, Map<Setting<Integer>, Integer> se
     public static final String SETTING_VERSION_UPGRADED = "index.version.upgraded";
     public static final String SETTING_VERSION_UPGRADED_STRING = "index.version.upgraded_string";
     public static final String SETTING_CREATION_DATE = "index.creation_date";
+    public static final String SETTING_LIFECYCLE_ORIGINATION_DATE = "index.lifecycle.origination_date";
```
This setting should go within the `x-pack` plugin code rather than in the server code, since it's related to the ILM plugin. I would recommend putting it in `LifecycleSettings` since the other ILM-specific settings go there.

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -217,6 +219,7 @@ public void validate(Integer numRoutingShards, Map<Setting<Integer>, Integer> se
     public static final Setting<Integer> INDEX_PRIORITY_SETTING =
         Setting.intSetting("index.priority", 1, 0, Property.Dynamic, Property.IndexScope);
     public static final String SETTING_CREATION_DATE_STRING = "index.creation_date_string";
+    public static final String SETTING_LIFECYLE_ORIGINATION_DATE_STRING = "index.lifecyle.origination_date_string";
```
Same here about moving to `LifecycleSettings`

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -954,6 +961,11 @@ public Builder creationDate(long creationDate) {
             return this;
         }
 
+        public Builder lifecycleOriginationDate(long originationDate) {
```
I don't think we need the equivalent to this in x-pack, `IndexMetaData` has settings that are considered "core" to an index (we don't have builders for every one of them)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetaData.java:
```diff
@@ -1440,6 +1452,11 @@ public static Settings addHumanReadableSettings(Settings settings) {
             ZonedDateTime creationDateTime = ZonedDateTime.ofInstant(Instant.ofEpochMilli(creationDate), ZoneOffset.UTC);
             builder.put(SETTING_CREATION_DATE_STRING, creationDateTime.toString());
         }
+        Long lifecycleOriginationDate = settings.getAsLong(SETTING_LIFECYCLE_ORIGINATION_DATE, null);
```
This code is also probably not needed since we don't want x-pack settings in the `server` module

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleExecutionState.java:
```diff
@@ -310,8 +321,14 @@ public Builder setStepTime(Long stepTime) {
             return this;
         }
 
+        public Builder setIndexOriginationDate(Long originationDate) {
```
I think we should approach this slightly differently, rather than storing the origination date in the `LifecycleExecutionState`.

I think it should be a regular index setting (keeping the original creation date in the `LifecycleExecutionState`), then, when we go to use the creation date, we retrieve the setting (if set) and use that rather than the date stored in the execution state. This has the benefit of allowing a user to update the `index.lifecycle.origination_date` setting after the index has already been created, or remove the setting after previously overriding the date (we can fall back to the original creation date).

andreidan replied:
Ah yes, great point

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleExecutionState.java:
```diff
@@ -310,8 +321,14 @@ public Builder setStepTime(Long stepTime) {
         public LifecycleExecutionState build() {
-            return new LifecycleExecutionState(phase, action, step, failedStep, stepInfo, phaseDefinition, indexCreationDate,
+            return new LifecycleExecutionState(phase, action, step, failedStep, stepInfo, phaseDefinition,
+                indexOriginationDate != null && indexOriginationDate != -1L ? indexOriginationDate : indexCreationDate,
```
See my above comment about not overwriting the `indexCreationDate`, if we only store one then a user can't go back to the "default" value for the date.

dakrone reviewed (changes requested):
I left some comments on this, I think we should approach it more towards the "read time" rather than overwriting the creation date when initializing the policy.

dakrone commented:
@andreidan can you also add version labels to this PR? Since this is an enhancement it should go to 8.0.0 and 7.5.0 (latest master and 7.x branch versions)

andreidan committed:
[ILM] Don't override creation date on policy init

The initial approach we took was to override the lifecycle creation date
if the `index.lifecycle.origination_date` setting was set. This had the
disadvantage of the user not being able to update the `origination_date`
anymore once set.

This commit changes the way we makes use of the
`index.lifecycle.origination_date` setting by checking its value when
we calculate the index age (ie. at "read time") and, in case it's not
set, default to the index creation date.

andreidan commented:
retest this please

dakrone commented on docs/reference/ilm/policy-definitions.asciidoc:
```diff
@@ -60,10 +60,12 @@ phase after one day. Until then, the index is in a waiting state. After
+case the `min_age` will be the time elapsed since that specified date. If the
+index is rolled over, then `min_age` is the time elapsed from the time the
+index is rolled over. The intention here is to execute following phases and
+actions relative to when data was written last to a rolled over index.
```
I think this documentation looks good, but can you also add some in `ilm-settings.asciidoc` for the `index.lifecycle.origination_date` setting

dakrone commented on x-pack/plugin/ilm/src/test/java/org/elasticsearch/xpack/ilm/IndexLifecycleInitialisationTests.java:
```diff
@@ -201,7 +201,14 @@ public void testExplainExecution() throws Exception {
+                createIndexRequest("test").settings(
+                    Settings.builder()
+                        .put(settings)
+                        .put(LifecycleSettings.SETTING_LIFECYCLE_ORIGINATION_DATE, 1000L)
```
Can you add another check underneath this that setting this setting back to `null` goes back to using the original lifecycle creation date?

dakrone reviewed (commented):
This approach looks much better, thanks @andreidan, I left some more comments that should be addressed, but nothing major

dakrone commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/LifecycleSettings.java:
```diff
@@ -29,6 +30,8 @@
     public static final Setting<Boolean> LIFECYCLE_INDEXING_COMPLETE_SETTING = Setting.boolSetting(LIFECYCLE_INDEXING_COMPLETE, false,
         Setting.Property.Dynamic, Setting.Property.IndexScope);
+    public static final Setting<Long> LIFECYCLE_ORIGINATION_DATE_SETTING =
+        Setting.longSetting(SETTING_LIFECYCLE_ORIGINATION_DATE, -1, -1, Setting.Property.IndexScope, Setting.Property.NodeScope);
```
Good catch :)

andreidan committed:
Drop empty line

andreidan committed:
Undo constructor call arguments wrap

andreidan committed:
Drop `SETTING` from variable name

andreidan committed:
Make origination date setting index scope dynamic

andreidan committed:
Document orignation date setting in ilm settings

andreidan committed:
Test setting the origination date back to null

andreidan committed:
Drop quotes from setting description

dakrone reviewed (approved):
LGTM, thanks for iterating on this @andreidan

elasticmachine committed:
Merge branch 'master' into ilm-origination-date
